tests/std/tests/P0896R4_ranges_subrange/test.cpp

```diff
@@ -0,0 +1,1223 @@
+
+    template <class Rng>
+    constexpr bool test_construction() {
+        STATIC_ASSERT(same_as<Rng, std::remove_cvref_t<Rng>>);
+
+        using I = ranges::iterator_t<Rng>;
+        using S = ranges::sentinel_t<Rng>;
+
+        STATIC_ASSERT(test_subrange<subrange<I, S, subrange_kind::sized>, Rng>());
+        if constexpr (!sized_sentinel_for<S, I>) {
+            STATIC_ASSERT(test_subrange<subrange<I, S, subrange_kind::unsized>, Rng>());
+        }
+
+        return true;
+    }
+    STATIC_ASSERT(test_construction<test_range<output_iterator_tag, int, false, false>>());
+    STATIC_ASSERT(test_construction<test_range<output_iterator_tag, int, true, false>>());
+    STATIC_ASSERT(test_construction<test_range<input_iterator_tag, int, false, false>>());
+    STATIC_ASSERT(test_construction<test_range<input_iterator_tag, int, true, false>>());
+    STATIC_ASSERT(test_construction<test_range<forward_iterator_tag, int, false, false>>());
+    STATIC_ASSERT(test_construction<test_range<forward_iterator_tag, int, false, true>>());
+    STATIC_ASSERT(test_construction<test_range<forward_iterator_tag, int, true, false>>());
+    STATIC_ASSERT(test_construction<test_range<forward_iterator_tag, int, true, true>>());
+    STATIC_ASSERT(test_construction<test_range<bidirectional_iterator_tag, int, false, false>>());
+    STATIC_ASSERT(test_construction<test_range<bidirectional_iterator_tag, int, false, true>>());
+    STATIC_ASSERT(test_construction<test_range<bidirectional_iterator_tag, int, true, false>>());
+    STATIC_ASSERT(test_construction<test_range<bidirectional_iterator_tag, int, true, true>>());
+    STATIC_ASSERT(test_construction<test_range<random_access_iterator_tag, int, true, false>>());
+    STATIC_ASSERT(test_construction<test_range<random_access_iterator_tag, int, true, true>>());
+    STATIC_ASSERT(test_construction<test_range<contiguous_iterator_tag, int, true, false>>());
+    STATIC_ASSERT(test_construction<test_range<contiguous_iterator_tag, int, true, true>>());
+
+    STATIC_ASSERT(test_construction<std::forward_list<int>>());
+    STATIC_ASSERT(test_construction<std::list<int>>());
+    STATIC_ASSERT(test_construction<std::vector<int>>());
+    STATIC_ASSERT(test_construction<std::string_view>());
+
+    // Validate that slicing conversions are forbidden
+    struct Base {};
+    struct Derived : Base {};
+    STATIC_ASSERT(!std::constructible_from<subrange<Base*>, Derived*, Derived*>);
+    STATIC_ASSERT(!std::constructible_from<subrange<Base*>, Derived*, Derived*, std::make_unsigned_t<std::ptrdiff_t>>);
+    STATIC_ASSERT(!std::constructible_from<subrange<Base*>, subrange<Derived*>>);
+    STATIC_ASSERT(!std::constructible_from<subrange<Base*>, subrange<Derived*>, std::make_unsigned_t<std::ptrdiff_t>>);
+
+    struct with_converting_iterators {
+        template <bool IsConst>
+        struct iterator {
+            using iterator_concept  = input_iterator_tag;
+            using iterator_category = output_iterator_tag;

```

Comment:
Is this intentional?

Reply:
Somewhat - this is a move-only single-pass C++20-but-not-C++17 iterator (LWG-3289). We have two other similar occurrences, only one of which has a `// TRANSITION, LWG-3289` comment 😿. 

My hunch is that the fix for LWG-3289 is to allow `void` as an `iterator_category` to indicate "I am no kind of C++17 iterator". All of the iterator and range adaptors in the Standard will need fixing, but I think that change works correctly with the rest of the library as currently specified. As a test and to defend against regression, I'll change our three occurrences to `using iterator_category = void; // TRANSITION, LWG-3289` until and unless I need something different for view adaptor tests.